File: linked-lists/intersection-of-two-linked-lists/intersection-of-two-linked-lists.go
package intersectionoftwolinkedlists

type ListNode struct {
	Val  int
	Next *ListNode
}

func getIntersectionNode(headA, headB *ListNode) *ListNode {
	if headA == nil || headB == nil {
		return nil
	}

	aPointer := headA
	bPointer := headB

	for aPointer != bPointer {
		aPointer = aPointer.Next
		bPointer = bPointer.Next

		if aPointer == bPointer {
			return aPointer
		}

		if aPointer == nil {
			aPointer = headB
		}

		if bPointer == nil {
			bPointer = headA
		}
	}

	return aPointer
}

File: linked-lists/intersection-of-two-linked-lists/intersection-of-two-linked-lists.ts
function getIntersectionNode(headA: ListNode | null, headB: ListNode | null): ListNode | null {
  if (headA === null || headB === null) {
    return null;
  }

  let aPointer: ListNode | null = headA;
  let bPointer: ListNode | null = headB;

  while (aPointer != bPointer) {
    aPointer = aPointer.next;
    bPointer = bPointer.next;

    if (aPointer == bPointer) {
      return aPointer;
    }

    if (aPointer == null) {
      aPointer = headB;
    }

    if (bPointer == null) {
      bPointer = headA;
    }
  }

  return aPointer;
}
